### PR TITLE
Kook-1550 Clean up timer logic 

### PR DESCRIPTION
**Pull request by Jese310S**
JIRA: 🎫  [KOOK-1550](https://zendesk.atlassian.net/browse/KOOK-1550)

### Context
In order to resolve [KOOK-912](https://zendesk.atlassian.net/browse/KOOK-912) team kookaburra debugged to understand what may be the best solution on how to utilise the post message that` zaf_sdk` now sends to `lotus`.  It was realised that we do not need a timer in `zaf_sdk,` we only need the post message to send every time there is an event in an app.  Then `lotus` will handle any timer needed.  This  PR attempts to clean up the timer logic from `idleState.ts` and `client.js`.  The PR also adds` lodash debounce` to minimise the events to run once every 1000ms.  With these changes to the `zaf_sdk` Kookaburra can focus on listening for the post message in lotus to extend the session and fix the bug causing sessions to timeout. 

## PR
-cleans up timer logic
-adds debounce

## Screen Recording 

https://user-images.githubusercontent.com/59402195/176564562-ef46eb1d-d3be-4f08-a7d8-1f3d3e7be51b.mov

/cc @zendesk/vegemite

### DevQA Steps

- Make sure [Zendesk Apps Framework regression tests](https://zendesk.atlassian.net/wiki/spaces/ENG/pages/641702848/DevQA+process+in+Vegemite#DevQAprocessinVegemite-regression-checks) are passing with this change.

NOTE: DevQA steps are to be actioned only once code has been reviewed and approved.

### Risks
* [HIGH | medium | low] Does it work across browsers (including IE!)?
* [HIGH | medium | low] Does it work in the different products (Support, Chat, Connect)?
* [HIGH | medium | low] Are there any performance implications?
* [HIGH | medium | low] Any security risks?
* [HIGH | medium | low] What features does this touch?

### Rollback Plan

1. Quickly [roll back] to the prior release so that customers can refresh to resolve their issue.
1. Revert this PR to restore the master branch to a deployable green state.
1. Notify the author.

[roll back]: https://github.com/zendesk/zendesk_app_framework_sdk/blob/master/DEPLOY.md#recovery

### CODE REVIEW

File: lib/client.js
Comment (josh-mo) [inline lib/client.js:355]: Think you can also remove the `data.iframe_session_timeout`

File: lib/utils/idleState.ts
Comment (josh-mo) [inline lib/utils/idleState.ts:88]: The `timeout` might be removed

File: lib/utils/idleState.ts
Comment (josh-mo) [inline lib/utils/idleState.ts:89]: If you are keen on a deep clean-up, I guess you can also remove the `stateProperty`.

File: lib/client.js
Comment (Jese310S) [reply lib/client.js:355]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅   Thanks for the review Josh 🙇‍♀️ .  The changes have been made.

File: lib/utils/idleState.ts
Comment (damenturnbull) [inline lib/utils/idleState.ts:70]: I think the `timeout` property can be cleaned up, given it's been removed from the constructor.
```suggestion
```

File: lib/client.js
Comment (damenturnbull) [inline lib/client.js:358]: Could I interest you in adding the debounce here instead? It would mean the IdleState class doesn't have to care about debouncing, and it can just call markActive whenever an event fires. ie. it makes debouncing a `client` concern, couples debouncing and post message logic together, and makes the intent a little clearer.
```suggestion
      this._idleState.addObserver(debounce(() => {
        this.postMessage('session.live')
        console.log('** post message sent: session.live')
      }, 1000))
```

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts:115]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅

File: lib/client.js
Comment (Jese310S) [reply lib/client.js]: ✅

File: lib/utils/idleState.ts
Comment (damenturnbull) [inline lib/utils/idleState.ts:11]: Can `timer` be removed?
```suggestion
```

File: lib/client.js
Comment (damenturnbull) [inline lib/client.js:357]: A comment explaining why we used a debounce may be helpful here:
```suggestion
        /**
         * Sends an indication that the user is active in the iframe.
         * Adding a debounce avoids a bombardment of unnecessary postMessages every microsecond (like when a user drags their mouse).
         */
        this.postMessage('session.live')
```

File: lib/utils/idleState.ts
Comment (token-cjg) [inline lib/utils/idleState.ts:3]: You can probably delete these constants as I suspect they are no longer being used.

**Comment by token-cjg** [comment]
Not critical, but it would be valuable to add some tests to `lib/utils/idleState.ts`, `lib/utils/listener.ts`, and also just for `lib/client.js` as well - particularly considering the criticality of this asset.

Otherwise LGTM, I'll let one or two other Vegemites have a careful look first though before dropping an approval.

File: lib/utils/idleState.ts
Comment (Jese310S) [reply lib/utils/idleState.ts]: ✅   done, thank you 🙇‍♀️

**Review by etanxing — APPROVED** [review]

File: lib/utils/idleState.ts
Comment (etanxing) [inline lib/utils/idleState.ts:53]: Is `null` necessary here?

**Review by token-cjg — APPROVED** [review]

**Review by token-cjg — APPROVED** [review]

**Review by adrianbonnici — APPROVED** [review]
Great work on simplifying the implementation 🎉

**Review by damenturnbull — APPROVED** [review]
Thanks for adding unit tests @Jese310S. Here's another 👍 to add to your collection of approvals... 😄

**Comment by Jese310S** [comment]
> 

@token-cjg we have finished adding unit tests for `lib/utils/idleState.ts`and  `lib/utils/listener.ts` .  This PR should now be ready for Vegemite to test and release.  🙇‍♀️   thanks

**Comment by token-cjg** [comment]
cc @zach-anthony @txu-gsd

**Pull request merged**
